articles/quadratic-arithmetic- programs-from-zero-to-hero/quadratic_arithmetic_programs_from_zero_to_hero.md

```diff
@@ -0,0 +1,149 @@
+
+![Pipeline image](./attachments/pipeline.png)
+The steps here can be broken up into two halves. First, zk-SNARKs cannot be applied to any computational problem directly; rather, you have to convert the problem into the right “form” for the problem to operate on. The form is called a “quadratic arithmetic program” (QAP).
+
+We will use a common-used zk-SNARK example  of equation $x^3 + x + 5 = 35$ here. The prover need to prove he know the answer to this equation (which is $x = 3$)
+# Flattening
+
+Break the equation down into a sequence of statement that is of two form:
+
+$x = y$ (where y can be a variable or a number)
+
+or 
+
+$x = y$ (op) $z$ (where op can be +, -, *, / and y and z can be variables, numbers or sub-expressions)
+
+You can think of those as logic gates in a circuit.
+
+The result of the flattening process for the above equation is as follows:
+
+$sym_1 = x * x$
+
+$y = sym_1 * x$
+
+$sym_2 = y + x$
+
+$~out = sym_2 + 5$
+
+# R1CS
+
+R1CS (Rank-1 Constrain System) is a system where you turn every equation from Flattening step to a constrain system with the following formula
+
+> $A . s * B . s - C . s = 0$
+
+where A, B, C is a vector represent the mapping for the variables
+s is the full variable list 
+
+Firstly, you take the result from flattening step 
+which look like this
+
+$C = A * B$
+
+and change it to this form
+
+$A * B = C$
+
+then 
+
+$A * B - C = 0$
+
+so we will have
+
+$x * x - sym_1 = 0$
+
+$x * sym_1 - y = 0$
+
+$(x+y)*1 = sym_2$
+
+$(sym_2 + 5)*1 - ~out = 0$
+
+then we will map all 4 line into vector A, B and C corresponding to the format above using this mapping 
+
+['~one', 'x', '~out', 'sym_1', 'y', 'sym_2']
+
+we will have mapping vector A, B, C and variable vector s as follows:
+
+A  
+[0, 1, 0, 0, 0, 0]  
+[0, 0, 0, 1, 0, 0]  
+[0, 1, 0, 0, 1, 0]  
+[5, 0, 0, 0, 0, 1]
+
+B  
+[0, 1, 0, 0, 0, 0]  
+[0, 1, 0, 0, 0, 0]  
+[1, 0, 0, 0, 0, 0]  
+[1, 0, 0, 0, 0, 0]
+
+C  
+[0, 0, 0, 1, 0, 0]  
+[0, 0, 0, 0, 1, 0]  
+[0, 0, 0, 0, 0, 1]  
+[0, 0, 1, 0, 0, 0]
+
+s
+[1, 3, 35, 9, 27, 30]
+
+so you might be asking how do we get  $A . s * B . s - C . s = 0$ to be true. So it all come to the property of dot product.
+
+For example:
+
+if you have a vector [1, 2, 3, 4, 5, 6]
+
+then dot product it with vector [0, 0, 1, 0, 0, 0]
+
+we have 3 as the result
+
+
+Apply that logic to the first line of A, B and C we have 
+
+$x * x - sym_1 = 0$
+
+so it mean that we basically rewrite the equation, but in a matrix form.
+
+# QAP
+
+Turn all columns of A 
+into a matrix of polynomial f(x) of n-1 degree using _Lagrange interpolation_ 
+
+For example: 
+
+the first column of matrix A is 
+
+0 0 0 5
+
+so we will use _Lagrange interpolation_ to find a polynomial of 3 degree which equal to those 4 values at x=1, x=2, x=3 and x=4 or we can say that the polynomial go through 4 points (1, 0), (2, 0), (3, 0), (4, 5) 
```

Comment:
Link to our definition of Lagrange interpolation

Reply:
new term file created